### ISSUE TITLE
User preferences are not supported

### ISSUE THREAD

**Issue by dewyatt** [riboseinc/rnp#36]
Each User ID can have preferred symmetric algs, hash algs, compression algs, key server prefs, etc.
Similar to key flags, I think this is all parsed but not utilized and can not be written out either.

It may be helpful to expand the definition of a "User ID" to include preferences. That would be similar to how PGPy (python) handles it, for example:

```python
key.add_uid(uid, usage={KeyFlags.Sign, KeyFlags.EncryptCommunications, KeyFlags.EncryptStorage},
            hashes=[HashAlgorithm.SHA256, HashAlgorithm.SHA384, HashAlgorithm.SHA512, HashAlgorithm.SHA224],
            ciphers=[SymmetricKeyAlgorithm.AES256, SymmetricKeyAlgorithm.AES192, SymmetricKeyAlgorithm.AES128],
            compression=[CompressionAlgorithm.ZLIB, CompressionAlgorithm.BZ2, CompressionAlgorithm.ZIP, CompressionAlgorithm.Uncompressed])
```

**Comment by ronaldtse** [comment]
Agreed. We should ensure that these preferences can be written out (and generated) properly.

User ID, Preferred Algorithms and Key Flags are all sub-packets of a Signature packet, so it's not strange to have them together. 

However, Preferred Algorithms and Key Flags are used mainly in self-signature packets (Key Flags can also be used in "certification signatures"). Would a "self-signature" construct easier support this?

**Comment by dewyatt** [comment]
We do now support loading and saving user preferences.

So what remains on this issue is to actually utilize these preferences.

I think this has potential to be tricky in some cases. For example, if you were signing+encrypting, you may want to consider all of:

* the recipient's preferences
* the sender's preferences
* the algorithm requirements (like required digest sizes for the PK alg, etc).

**Comment by dewyatt** [comment]
> Would a "self-signature" construct easier support this?

I think the oddly-named `pgp_subsig_t` is the closest thing we have to this at the moment. It contains *all* signatures that occur in a keyring. So it's more than just self-sigs. It could probably be split out with a bit of work. It could potentially be split further than that since self-sigs include certifications, subkey bindings, and direct-key sigs.

**Comment by ronaldtse** [comment]
@dewyatt is this fixed? Thanks!

**Comment by ni4** [comment]
Closing this as completed as we have full support of API for most of signature subpackets/preferences. If something is missing then we could create additional issue.
